pkg/apis/certmanager/v1alpha1/types_issuer.go

```diff
@@ -204,15 +204,104 @@ type ACMEIssuer struct {
 	// user account.
 	PrivateKey SecretKeySelector `json:"privateKeySecretRef"`
 
-	// HTTP-01 config
+	// Solvers is a list of challenge solvers that will be used to solve
+	// ACME challenges for the matching domains.
+	// +optional
+	Solvers []ACMEChallengeSolver `json:"solvers,omitempty"`
+
+	// DEPRECATED: HTTP-01 config
 	// +optional
 	HTTP01 *ACMEIssuerHTTP01Config `json:"http01,omitempty"`
 
-	// DNS-01 config
+	// DEPRECATED: DNS-01 config
 	// +optional
 	DNS01 *ACMEIssuerDNS01Config `json:"dns01,omitempty"`
 }
 
+type ACMEChallengeSolver struct {
+	// Selector selects a set of DNSNames on the Certificate resource that
+	// should be solved using this challenge solver.
+	Selector *CertificateDNSNameSelector `json:"selector,omitempty"`
+
+	// +optional
+	HTTP01 *ACMEChallengeSolverHTTP01 `json:"http01,omitempty"`
+
+	// +optional
+	DNS01 *ACMEChallengeSolverDNS01 `json:"dns01,omitempty"`
+}
+
+// CertificateDomainSelector selects certificates using a label selector, and
+// can optionally select individual DNS names within those certificates.
+// If both MatchLabels and DNSNames are empty, this selector will match all
+// certificates and DNS names within them.
+type CertificateDNSNameSelector struct {
+	// A label selector that is used to refine the set of certificate's that
+	// this challenge solver will apply to.
+	// TODO: use kubernetes standard types for matchLabels
+	// +optional
+	MatchLabels map[string]string `json:"matchLabels,omitempty"`
+
+	// List of DNSNames that can be used to further refine the domains that
+	// this solver applies to.
+	// +optional
+	DNSNames []string `json:"dnsNames,omitempty"`
+}
+
+type ACMEChallengeSolverHTTP01 struct {
+	// +optional
+	Ingress *ACMEChallengeSolverHTTP01Ingress `json:"ingress"`
+}
+
+type ACMEChallengeSolverHTTP01Ingress struct {
+	// Optional service type for Kubernetes solver service
+	// +optional
+	ServiceType corev1.ServiceType `json:"serviceType,omitempty"`
+
+	// +optional
+	IngressClass *string `json:"ingressClass,omitempty"`
+
+	// +optional
+	IngressName string `json:"ingressName,omitempty"`
```

Comment:
/cc @kragniz should these two fields be `Class` and `Name` to avoid the stutter?

Reply:
Updated to Name and Class 😄